core/src/main/java/io/questdb/cairo/TableWriterSegmentCopyInfo.java

```diff
@@ -0,0 +1,168 @@
+        txns.add(segmentRowOffset);
+        txns.add(relativeSeqTxn);
+        txns.add(committedRowsCount);
+        txns.add(segmentIndex);
+
+        if (seqTxnOrder.size() > 0) {
+            seqTxnOrder.set(relativeSeqTxn, (int) (txns.size() / 4 - 1));
+        }
+        maxTxnRowCount = Math.max(maxTxnRowCount, committedRowsCount);
+        totalRows += committedRowsCount;
+        this.minTimestamp = Math.min(this.minTimestamp, minTimestamp);
+        this.maxTimestamp = Math.max(this.maxTimestamp, maxTimestamp);
+    }
+
+    public void clear() {
+        segments.clear();
+        txns.clear();
+        seqTxnOrder.clear();
+        totalRows = 0;
+        maxTxnRowCount = 0;
+        startSeqTxn = 0;
+        minTimestamp = Long.MAX_VALUE;
+        maxTimestamp = Long.MIN_VALUE;
+        hasSegmentGap = false;
+        allDataInOrder = false;
+    }
+
+    @Override
+    public void close() {
+        segments = Misc.free(segments);
+        txns = Misc.free(txns);
+    }
+
+    public boolean getAllTxnDataInOrder() {
+        return allDataInOrder;
+    }
+
+    public int getMappingOrder(long absoluteSeqTxn) {
+        return seqTxnOrder.get((int) (absoluteSeqTxn - startSeqTxn));
+    }
+
+    public long getMaxTimestamp() {
+        return maxTimestamp;
+    }
+
+    public long getMaxTxRowCount() {
+        return maxTxnRowCount;
+    }
+
+    public long getMinTimestamp() {
+        return minTimestamp;
+    }
+
+    public long getRowHi(int segmentIndex) {
+        return Math.abs(segments.get(segmentIndex * 4L + 3));
+    }
+
+    public long getRowLo(int segmentIndex) {
+        return Math.abs(segments.get(segmentIndex * 4L + 2));
+    }
+
+    public int getSegmentCount() {
+        return (int) (segments.size() / 4);
+    }
+
+    public int getSegmentId(int segmentIndex) {
+        return (int) segments.get(segmentIndex * 4L + 1);
+    }
+
+    public long getSegmentsAddress() {
+        return segments.getAddress();
+    }
+
+    public long getStartTxn() {
+        return startSeqTxn;
+    }
+
+    public long getTotalRows() {
+        return totalRows;
+    }
+
+    public long getTxnCount() {
+        return txns.size() / 4;
+    }
+
+    public long getTxnInfoAddress() {
+        return txns.getAddress();
+    }
+
+    public int getWalId(int segmentIndex) {
+        return (int) segments.get(segmentIndex * 4L);
+    }
+
+    public boolean hasSegmentGaps() {
+        return hasSegmentGap;
+    }
+
+    public void initBlock(long startSeqTxn, int txnCount, boolean hasSymbols) {
+        this.startSeqTxn = startSeqTxn;
+        if (hasSymbols) {
+            seqTxnOrder.setPos(txnCount);
+        }
+    }
+
+    public boolean isLastSegmentUse(int segmentIndex) {
+        return segments.get(segmentIndex * 4L + 3) > 0;
+    }
+
+    public void setAllTxnDataInOrder(boolean allInOrder) {
+        this.allDataInOrder = allInOrder;
+    }
+
+    public void setSegmentGap(boolean value) {
+        hasSegmentGap = value;
```

Comment:
do we have tests to cover this? I could not find any in "cairo"

Reply:
No, it's only a theoretical possibility that a segment can have an uncommitted range of rows. If this ever happens, the new apply algo cannot be used, so we fall back to txn by txn approach.